### ISSUE TITLE
How shousl

### ISSUE THREAD

**Issue by fpagnoux** [openfisca/openfisca-core#638]
Hi there!

I really enjoy OpenFisca, but I recently encountered an issue.

## Context

I identify more as a:

- Analyst _(I make macroscopic computations on real populations)_.
- Business expert _(I create tests and model legislation)_.
- Commenter _(I make data visualisations)_.
- Developer _(I create tools that use the existing OpenFisca code)_.
- Historian _(I accumulate data based on past legislation values)_.
- Lobbyist _(I model reforms to make them exist)_.
- Producer _(I make computations on individual situations)_.

(remove this line as well as all items in the list that don't fit your situation)

**Comment by fpagnoux** [comment]
**@MattiSG's answer:

**4. Standard library** (as includable helpers)
**5. Patterns** (documented in a wiki)

**Comment by fpagnoux** [comment]
> What that case is really calling for, in my opinion, is a date diff helper, most probably provided by an external library.

Indeed, that would be worth exploring. But we may hit a wall: we are not looking for _any_ generic python date library, we are looking for a _numpy_ one 😈.

> I would need to get two other examples of cases where that question would apply

I think this is the only example where we'd like to share something to _all_ OpenFisca re-users. However, there have been more _model specific_ needs for formula reusable templates:
- The IPP needed a way to easily build [quantile variables](https://github.com/openfisca/openfisca-survey-manager/pull/47#pullrequestreview-100852797) (discussion is in French, but worth reading).
- In France, most of social cotisations are computed the same way. The current way of factorizing these similarities is quite messy, and templates have been mentioned several times as a cleaner solution.

**Comment by benjello** [comment]
I confirm @fpagnoux examples.
I can add another generic yet to be coded examples where it could be very useful.
It is convenient to compute marginal or average tax rate according to some revenue (like in [here]((https://github.com/openfisca/openfisca-core/blob/master/openfisca_core/rates.py)) at the individual level. This would involve computing for example the difference between the value of the tax paid when the revenue of _every individual_ varies by some amount without manually running two simulations.

**Comment by MattiSG** [comment]
Thanks for contributing these examples, I feel we're making a lot of progress 🙂

I definitely see the usefulness of sharing economic analysis functions. This would make it easier over time to have cross-country comparisons.
I see how we could, building upon that business need, also provide a technical utility function that would make date manipulation easier.

Can we elaborate a bit on this “templates” idea? I know the ins and outs of providing a standard library and can see quite deeply how we could collaborate on it, maintain and upgrade it, and make it modular. From what I understand, “templates” would be more like a set of pre-written functions that would then be stale and that you would need to copy-and-paste (or —gasp— `git subtree`) to update. Am I missing something in terms of implementation? It looks like you have already thought it through 🙂

**Comment by fpagnoux** [comment]
>  “templates” would be more like a set of pre-written functions that would then be stale and that you would need to copy-and-paste

Maybe templates was a bad choice of word : we want to avoid copy-pasting ;)

Several implementations suggestions have been mentioned: see suggestions [1](https://github.com/openfisca/openfisca-survey-manager/pull/47#discussion_r171912096#discussion_r172002762), [2](https://github.com/openfisca/openfisca-survey-manager/pull/47#discussion_r171912096#discussion_r172005844), and [2bis](https://github.com/openfisca/openfisca-survey-manager/pull/47#discussion_r171912096#discussion_r172006310).

This is for what a template could _be_. We have not discussed how a templates could be _shared_ across country packages yet, but we could imagine another openfisca-templates library, I guess.
